blogpost fido/passkeys (#845)
ordina-jworks/ordina-jworks.github.io: _posts/2024-01-16-fido-passkey.md

FIDO (Fast IDentity Online) is a set of open authentication standards that aims to replace traditional passwords with more secure and convenient methods.

Key benefits of FIDO authentication:
* Strong Security: FIDO Authentication utilises public-key cryptography, which is more secure than traditional passwords against phishing attacks, keyloggers, and other threats.
* Convenience: Passkeys, the credential type enabled by FIDO Authentication, are stored on your device and can be unlocked using biometrics like fingerprint or face recognition, eliminating the need to remember or type passwords repeatedly.

## How does it work

To understand how passkeys work, let's take a step back and delve into the realm of asymmetric cryptography. In asymmetric cryptography, 
we deal with a key pair consisting of a private key, which must remain private and undisclosed, and a public key, which can be shared openly.

Through asymmetric encryption, a message encrypted with the public key can only be deciphered by the possessor of the private key. 

> Conversely, if a message is encrypted with the private key, it can be read using the public key (signature).

Reviewer: ```suggestion
Conversely, if a message is encrypted with the private key, it can be decrypted using the public key. 
This method is used in signatures, to validate the identity of the sender of the information.
```

So, how is this technology employed in FIDO? During registration, your device generates a unique passkey, which contains a public key and a private key.
The public key is stored on the service provider's server, while the private key remains securely on your device (smartphone, tablet, or a dedicated FIDO security key).

> To register as a new user, we must send our public key to the server and have it validated. 

Reviewer: ```suggestion
To register as a new user, we must send our public key to the server for validation. 
```

The server will send us a challenge that we must sign using our private key. 
If the server can verify the signature using our public key, the registration is complete.

To authenticate a user, we send our username to the server, and the server will answer with a challenge encrypted with our public key.
On our device, we need to decrypt the challenge with our private key, solve the challenge, and return our response encrypted with our private key.

> After sending the encrypted response, the server is able to validate the response with our public key and check if the login succeeded.

Reviewer: ```suggestion
After sending the encrypted response, the server is able to validate the response using our public key.
If the challenge is correct, the server approves the login. 
```

> So, in this way, we no longer store the user's password, but only a public key that can be shared openly.

Reviewer: ```suggestion
In this way we no longer store the user passwords, but only their public key which can be shared openly.
```

> 

Reviewer: Maybe add a small diagram demonstrating the flow with messages going back and forth? Or link to a good YouTube video with decent visuals.

### Add the needed dependencies

```xml
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-web</artifactId>
</dependency>

<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-data-jpa</artifactId>
</dependency>

<dependency>
    <groupId>com.yubico</groupId>
    <artifactId>webauthn-server-core</artifactId>
    <version>2.5.0</version>
</dependency>

<dependency>
    <groupId>org.projectlombok</groupId>
    <artifactId>lombok</artifactId>
    <scope>provided</scope>
</dependency>

<dependency>
    <groupId>com.h2database</groupId>
    <artifactId>h2</artifactId>
</dependency>

<dependency>
    <groupId>org.flywaydb</groupId>
    <artifactId>flyway-core</artifactId>
</dependency>
```

### Database
We use Flyway to set up our database; we need to keep track of our users and their passkeys.

We need to save the following data for the user.

The username is the identifier specified by the user, and the user handle is the unique identifier generated by the system, used to streamline communication between the FIDO server and the authenticator.\

> The assertion contains information such as user identification, the authentication method used, and cryptographic signatures to ensure the authenticity of the claim. It is used to confirm that the user has been successfully authenticated and that the authenticator is indeed trustworthy. It is also rebuilt every time a user attempts to log in.\

Reviewer: Confusing -> rewrite.

> The public key JSON is used when creating a new public key for a user, typically during the registration process. It contains all the parameters and options necessary for generating a new public key for the respective user. The authenticator will use these options to generate the new key.\

Reviewer: Confusing -> rewrite.

Reviewer: Explain what is generated where and when. This is unclear now (the authenticator = server? doesn't generate a public key per user right? The "user"/"user-device" does that.

                .build();
        return rp;
    }

    @Bean
    public AuthenticatorSelectionCriteria authenticatorSelectionCriteria() {
        AuthenticatorAttachment authenticatorAttachment = AuthenticatorAttachment.CROSS_PLATFORM;
        ResidentKeyRequirement residentKeyRequirement = ResidentKeyRequirement.PREFERRED;
        UserVerificationRequirement userVerificationRequirement = UserVerificationRequirement.PREFERRED;

        return AuthenticatorSelectionCriteria.builder()
                .authenticatorAttachment(authenticatorAttachment)
                .residentKey(residentKeyRequirement)
                .userVerification(userVerificationRequirement)
                .build();
    }

    @Bean
    public List<PublicKeyCredentialParameters> publicKeyCredentialParameters() {
        List<PublicKeyCredentialParameters> pubKeyCredParams = new ArrayList<>();

        PublicKeyCredentialParameters param1 = PublicKeyCredentialParameters.ES256;
        PublicKeyCredentialParameters param2 = PublicKeyCredentialParameters.RS256;
        pubKeyCredParams.add(param1);
        pubKeyCredParams.add(param2);
        return pubKeyCredParams;
    }
}
```

### Create Extension Output

> The `ClientExtensionOutputs` refers to the output of client extensions during the WebAuthn process, such as creating a public key credential or authentication. This output may contain information specific to the used extensions.

Reviewer: Add reference to WebAuthn for more information

        return FinishAssertionOptions.builder()
                .request(assertionRequest)
                .response(publicKeyCredential)
                .build();
    }
}
```

### Login Controller
Create a controller `LoginController` with 2 endpoints, the `start login` and `verify login`.

```java
@AllArgsConstructor
@RestController
@RequestMapping("/login")
@CrossOrigin("http://localhost:4200")
public class LoginController {
    private final StartLoginService startLoginService;
    private final VerifyLoginService verifyLoginService;

    @PostMapping("/start")
    public ResponseEntity<LoginResponseResource> startLogin(@RequestBody LoginRequestResource resource) throws JsonProcessingException {
        return ResponseEntity.ok(startLoginService.startLogin(resource));
    }

    @PostMapping("/verify")
    public ResponseEntity<VerifyLoginResponseResource> verifyLogin(@RequestBody VerifyLoginRequestResource resource) throws Base64UrlException, IOException {
        return ResponseEntity.ok(verifyLoginService.verify(resource));
    }
}
```

> 

Reviewer: Since there is very little explanation about the code itself (specific lines), consider moving it out of the blogpost (github repo) and just reference (in a bullet list or small paragraphs) what the different components will do.

            <p>Experience a new era of online security with FIDO Passwordless Authentication. Say goodbye to the hassle of remembering and managing passwords. 
            Our cutting-edge technology ensures a seamless and secure login process for users, making online interactions effortless and worry-free.</p>
        </div>

        <div class="col-sm-12 col-md-4">
            <form>
                <div class="mb-3">
                    <label for="username" class="form-label">Username</label>
                    <input type="text" class="form-control" id="username" aria-describedby="usernameHelp"
                           [(ngModel)]="username" [ngModelOptions]="{standalone: true}">
                    <div id="usernameHelp" class="form-text">We'll never share your info with anyone else.</div>
                </div>
                <button type="submit" class="btn btn-primary" (click)="register()">Register</button>
                <button type="submit" class="btn btn-primary ms-2" (click)="login()">Login</button>
            </form>
        </div>
    </div>
</div>
```

Paste the following code in the `user.component.html` file.
```html
<p>Welcome</p>

> ```

Reviewer: Same here, move the angular code to a github repo and reference it here. Explain what the different components do and why they are needed.

## Conclusion
Embracing FIDO and Passkeys represents a leap forward in security and user-friendliness. 
By shifting away from traditional password-based authentication methods, users benefit from heightened security measures while enjoying a more seamless and user-friendly experience. 
The combination of FIDO's robust security protocols and the convenience of passkeys not only enhances protection against cyber threats but also simplifies the user authentication process, 
contributing to a more secure and user-centric online environment. Embracing these advancements not only aligns with the evolving landscape of digital security 
but also empowers users with a more efficient and trustworthy means of safeguarding their online identities. 

> It's a win-win that positions FIDO and passkeys as a compelling choice for the future of secure and user-friendly authentication.

Reviewer: Include who is already supporting it (PayPal, IOS, Google services, ...)

## Extra
If you're interested in exploring the implementation details, you can access the frontend and backend code on GitHub.
* [Backend](https://github.com/nicholasM95/passwordless-backend){:target="_blank" rel="noopener noreferrer"}
* [Frontend](https://github.com/nicholasM95/passwordless-frontend){:target="_blank" rel="noopener noreferrer"}

If you want to start the application locally, you can start the Docker containers and visit [the project locally](http://localhost:4200){:target="_blank" rel="noopener noreferrer"}.
```bash

> docker run -p 8080:8080 nicholas95/passwordless-backend:v1

Reviewer: Make sure to include in the readme that this is not production ready code (even if it is...) and that it's provided as is. 

And include a licence ;) Preferably MIT.
